sync/pooled_worker_pool.go

```diff
@@ -73,14 +73,43 @@ func (p *pooledWorkerPool) Go(work Work) {
 	workCh <- work
 }
 
-func (p *pooledWorkerPool) spawnWorker(workCh chan Work) {
+func (p *pooledWorkerPool) GoOrGrow(work Work) {
+	// Use time.Now() to avoid excessive synchronization
+	workChIdx := p.nowFn().UnixNano() % p.numShards
+	workCh := p.workChs[workChIdx]
+	select {
+	case workCh <- work:
+	default:
+		// If the queue for the worker we were assigned to is full,
+		// allocate a new goroutine to do the work and then
+		// assign it to be a temporary additional worker for the queue.
+		// This allows the worker pool to accommodate "bursts" of
+		// traffic. Also, it reduces the need for operators to tune the size
+		// of the pool for a given workload. If the pool is initially
+		// sized too small, it will eventually grow to accommodate the
+		// workload, and if the workload decreases the killWorkerProbability
+		// will slowly shrink the pool back down to its original size because
+		// workers created in this manner will not spawn their replacement
+		// before killing themselves.
```

Comment:
Should we set a higher probability to kill these goroutines perhaps?

Reply:
I think the issue with setting a higher probability is it limits how far the pool can grow. I guess that can be a good or bad thing depending on how you look at it, but I'd rather the pool take longer to shrink (as long as it does shrink) if it allows it to accommodate a wider variety of workloads

Reply:
Hm don't have to change the probability for all go routines, only the new ones created during the burst. That should work yea?